activation/handler.go

```diff
@@ -82,7 +109,48 @@ func NewHandler(
 		inProgress: make(map[types.ATXID][]chan error),
 	}
 
-	return h
+	for _, opt := range opts {
+		opt(h)
+	}
+	if err := verifyAtxVersions(h.versions); err != nil {
+		return nil, fmt.Errorf("invalid ATX versions configured (%v): %w", h.versions, err)
+	}
+
+	h.log.With().Info("atx handler created",
+		log.Array("supported ATX versions", log.ArrayMarshalerFunc(func(enc zapcore.ArrayEncoder) error {
+			for _, v := range h.versions {
+				enc.AppendString(fmt.Sprintf("v%v from epoch %d", v.AtxVersion, v.publish))
+			}
+			return nil
+		})))
+
+	return h, nil
+}
+
+func verifyAtxVersions(versions []atxVersion) error {
+	if len(versions) == 0 {
+		return errors.New("no ATX versions configured")
+	}
+	slices.SortFunc(versions, func(a, b atxVersion) int {
+		switch {
+		case a.publish < b.publish:
+			return -1
+		case a.publish > b.publish:
+			return 1
+		}
+		return 0
+	})
+	lastVersion := types.AtxV1
+	for _, v := range versions {
+		if v.AtxVersion < types.AtxV1 || v.AtxVersion > types.AtxVMAX {
+			return fmt.Errorf("ATX version: %v not in range [%v:%v]", v, types.AtxV1, types.AtxVMAX)
+		}
+		if v.AtxVersion < lastVersion {
+			return fmt.Errorf("cannot decrease ATX version from %v to %v", lastVersion, v.AtxVersion)
+		}
+		lastVersion = v.AtxVersion
+	}
```

Comment:
This configuration is not ambiguous. It will end up as V1 in epochs [0:10) and V3 in epochs [10: :infinity:). Such a case is covered by the test `"v2 since epoch 0"`.

Reply:
But when is version 2 then used? Shouldn't this config be considered invalid because it says that v1 is valid until epoch 9 and version 3 starting epoch 10? Also consider this config:

```json
"atx-versions": {
      "10": 2,
      "10": 3,
      "10": 4,
 }
```

Here again v1 is valid until epoch 9 and version 4 starting with epoch 10?

Reply:
Unfortunately, JSON spec doesn't enforce keys to be unique and implementations can choose how to parse it. Viper uses the `encoding/json` from the standard library to decode JSON into a map. As far as I understand, this implementation will pick the last duplicated key. In your example, it would be `"10": 4`, without ambiguity here. Unfortunately, it doesn't fail to parse, but we can do nothing about it AFAIR.